Simplified/NYPLReaderReadiumView.m

```diff
@@ -576,154 +600,105 @@ - (void)readiumInitialize
-                                 [NSString stringWithFormat:@"ReadiumSDK.reader.openBook(%@)",
-                                  [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]]];
-                                
-                                NYPLLOG(@"opened server book location");
-                                
-                              }]];
 
-    }
-      if ((currentLocationString == nil && serverLocationString == nil) ||
-          (currentLocationString != nil && serverLocationString == nil) ||
-          (currentLocationString != nil && [deviceIDString isEqualToString:[NYPLAccount sharedAccount].deviceID]))
-      {
-        self.postLastRead = YES;
-      }
-      else if ((currentLocationString == nil && serverLocationString != nil) ||
-               (![currentLocationString isEqualToString:serverLocationString]) ||
-               (currentLocationString == nil && [deviceIDString isEqualToString:[NYPLAccount sharedAccount].deviceID]))
-      {
-        [[NYPLRootTabBarController sharedController] safelyPresentViewController:alertController animated:YES completion:nil];
-      }
-      else
-      {
-        self.postLastRead = YES;
-      }
-    
-  }];
+- (NSString*) currentChapter
+{
+  NYPLBookRegistry *registry = [NYPLBookRegistry sharedRegistry];
+  NYPLBookLocation *location = [registry locationForIdentifier:self.book.identifier];
+  NSDictionary *const locationDictionary = NYPLJSONObjectFromData([location.locationString dataUsingEncoding:NSUTF8StringEncoding]);
+  NSString *idref = locationDictionary[@"idref"];
+  return self.bookMapDictionary[idref][@"tocElementTitle"];
 }
 
-- (void) hasBookmarkForSpineItem:(NSString*)idref
+- (void)addBookmark
 {
-  
-  //  filter the bookmarks first by spine item (idref) and then run the result through a loop until there is a match
-  // break out the loop if a match was found 
-  
-  
-//  for (bookmark in filteredBookmarks) {
-  
+  NYPLBookRegistry *registry = [NYPLBookRegistry sharedRegistry];
+  NYPLBookLocation *location = [registry locationForIdentifier:self.book.identifier];
+  NSDictionary *const locationDictionary = NYPLJSONObjectFromData([location.locationString dataUsingEncoding:NSUTF8StringEncoding]);
+  NSString *contentCFI = locationDictionary[@"contentCFI"];
+  NSString *idref = locationDictionary[@"idref"];
+  NSString *chapter = self.bookMapDictionary[idref][@"tocElementTitle"];
 
-  // dummy cfi , replace by cfi from the bookmark in this loop 
-  NSString *contentCFI = @"/4/2/2";
-  
-  NSString *js = [NSString stringWithFormat:@"ReadiumSDK.reader.isVisibleSpineItemElementCfi('%@', '%@')",
-                  idref,
-                  contentCFI];
-  
-  NYPLLOG(js);
+  float progressWithinChapter = 0.0;
+  if (self.spineItemPageIndex > 0 && self.spineItemPageCount > 0) {
+    progressWithinChapter = (float) self.spineItemPageIndex / (float) self.spineItemPageCount;
+  }
+
+  NYPLReaderBookmark *bookmark = [[NYPLReaderBookmark alloc]
```

Comment:
@winniequinn  Things like the device ID and timestamp were typically set at certain times after initiazliation, but it seemed better to me to more strictly enforce parameters during init for fewer issues later.. and I audited most of the implicitly unwrapped optionals in the swift classes.